modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/time/ClusterTimeImpl.java

```diff
@@ -0,0 +1,172 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.metastorage.server.time;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.hlc.HybridClock;
+import org.apache.ignite.internal.hlc.HybridTimestamp;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.metastorage.impl.MetaStorageServiceImpl;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.internal.util.PendingComparableValuesTracker;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.TestOnly;
+
+/**
+ * Cluster time implementation with additional methods to adjust time and update safe time.
+ */
+public class ClusterTimeImpl implements ClusterTime {
+    /** The logger. */
+    private static final IgniteLogger LOG = Loggers.forClass(ClusterTimeImpl.class);
+
+    private final IgniteSpinBusyLock busyLock;
+
+    private volatile @Nullable LeaderTimer leaderTimer;
+
+    private final HybridClock clock;
+
+    private final PendingComparableValuesTracker<HybridTimestamp> safeTime;
+
+    /**
+     * Constructor.
+     *
+     * @param busyLock Busy lock.
+     * @param clock Node's hybrid clock.
+     */
+    public ClusterTimeImpl(IgniteSpinBusyLock busyLock, HybridClock clock) {
+        this.busyLock = busyLock;
+        this.clock = clock;
+        this.safeTime = new PendingComparableValuesTracker<>(clock.now());
+    }
+
+    /**
+     * Starts sync time scheduler.
+     *
+     * @param service MetaStorage service that is used by scheduler to sync time.
+     */
+    public void startLeaderTimer(MetaStorageServiceImpl service) {
+        if (!busyLock.enterBusy()) {
+            return;
+        }
+
+        try {
+            assert leaderTimer == null;
+
+            LeaderTimer newTimer = new LeaderTimer(service);
+
+            leaderTimer = newTimer;
+
+            newTimer.start();
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Stops sync time scheduler.
+     */
+    public void stopLeaderTimer() {
+        LeaderTimer timer = leaderTimer;
+
+        assert timer != null;
+
+        timer.stop();
+
+        leaderTimer = null;
+    }
+
+    @Override
+    public HybridTimestamp now() {
+        return clock.now();
+    }
+
+    @Override
+    public CompletableFuture<Void> waitFor(HybridTimestamp time) {
+        return safeTime.waitFor(time);
+    }
+
+    public void updateSafeTime(HybridTimestamp ts) {
+        this.safeTime.update(ts);
+    }
+
+    public void adjust(HybridTimestamp ts) {
```

Comment:
Little confuse name

Reply:
Why? We call adjusting the clock "adjust"

Reply:
Please add in the description of the method why it is.